src/d3d9/d3d9_device.cpp

```diff
@@ -4314,15 +4328,30 @@ namespace dxvk {
       const bool skipWait = (Flags & D3DLOCK_NOOVERWRITE) ||
                             quickRead                     ||
                             (boundsCheck && !pResource->DirtyRange().Overlaps(pResource->LockRange()));
-
       if (!skipWait) {
-        if (!(Flags & D3DLOCK_DONOTWAIT)) {
-          pResource->SetReadLocked(false);
-          pResource->DirtyRange().Clear();
-        }
+        if (desc.Pool & D3DPOOL_MANAGED) {
+          if (!WaitForResource(mappingBuffer, D3DLOCK_DONOTWAIT)) {
+            // if the mapped buffer is currently being copied to image and it's small enough
+            // we can just avoid a stall by allocating a new slice and copying the existing contents
+            DxvkBufferSliceHandle oldSlice = physSlice;
+            physSlice = pResource->DiscardMapSlice();
+            std::memcpy(physSlice.mapPtr, oldSlice.mapPtr, oldSlice.length);
+            EmitCs([
+              cBuffer = std::move(mappingBuffer),
+              cBufferSlice = physSlice
+            ] (DxvkContext* ctx) {
+              ctx->invalidateBuffer(cBuffer, cBufferSlice);
+            });
+          }
+        } else {
+          if (!(Flags & D3DLOCK_DONOTWAIT)) {
+            pResource->SetReadLocked(false);
```

Comment:
This needs a change but I can do that after the fact...

Reply:
What do you mean?

Reply:
this logic needs to be changed slightly, it can be made better + i think it should trigger in the other path?

Reply:
nvm ill fix this when it gets merged in

Reply:
actually nah add this to the other path please for now.